core/store/src/trie/state_parts.rs

```diff
@@ -46,60 +99,37 @@ impl Trie {
     /// Creating a StatePart takes all these nodes, validating a StatePart checks that it has the
     /// right set of nodes.
     fn visit_nodes_for_state_part(&self, part_id: PartId) -> Result<(), StorageError> {
-        let path_begin = self.find_path_for_part_boundary(part_id.idx, part_id.total)?;
-        let path_end = self.find_path_for_part_boundary(part_id.idx + 1, part_id.total)?;
+        let path_begin = self.find_state_part_boundary(part_id.idx, part_id.total)?;
+        let path_end = self.find_state_part_boundary(part_id.idx + 1, part_id.total)?;
 
         let mut iterator = self.iter()?;
         let nodes_list = iterator.visit_nodes_interval(&path_begin, &path_end)?;
         tracing::debug!(
             target: "state_parts",
             num_nodes = nodes_list.len());
 
-        // Extra nodes for compatibility with the previous version of computing state parts
-        if part_id.idx + 1 != part_id.total {
-            let mut iterator = self.iter()?;
-            let path_end_encoded = NibbleSlice::encode_nibbles(&path_end, false);
-            iterator
-                .seek_nibble_slice(NibbleSlice::from_encoded(&path_end_encoded[..]).0, false)?;
-            if let Some(item) = iterator.next() {
-                item?;
-            }
-        }
-
         Ok(())
     }
 
-    /// Part part_id has nodes with paths `[path(part_id), path(part_id + 1))`
-    /// path is returned as nibbles, last path is `vec![16]`, previous paths end
-    /// in nodes
-    pub fn find_path_for_part_boundary(
-        &self,
-        part_id: u64,
-        num_parts: u64,
-    ) -> Result<Vec<u8>, StorageError> {
-        assert!(part_id <= num_parts);
-        if part_id == num_parts {
-            return Ok(vec![16]);
-        }
-        let root_node = self.retrieve_node(&self.root)?.1;
-        let total_size = root_node.memory_usage;
-        let size_start = (total_size + num_parts - 1) / num_parts * part_id;
-        self.find_path(&root_node, size_start)
-    }
-
-    fn find_child(
+    /// Helper method for `find_node_in_dfs_order`.
+    /// Tells a child in which we should go to find a node in dfs order corresponding
+    /// to `memory_threshold`.
+    /// Accumulates `memory_skipped` as memory used by all skipped nodes.
+    /// Returns false if we already find desired node and should stop the process.
+    fn find_child_in_dfs_order(
         &self,
-        size_start: u64,
+        memory_threshold: u64,
         node: &mut TrieNodeWithSize,
-        size_skipped: &mut u64,
+        memory_skipped: &mut u64,
         key_nibbles: &mut Vec<u8>,
     ) -> Result<bool, StorageError> {
         let node_size = node.node.memory_usage_direct_no_memory();
-        if *size_skipped + node_size <= size_start {
-            *size_skipped += node_size;
-        } else {
+        if *memory_skipped + node_size <= memory_threshold {
+            *memory_skipped += node_size;
+        } else if node.node.has_value() {
```

Comment:
Don't we need to extend `key_nibbles` in case the current node is a `Leaf`, which is done at line 137?

Reply:
I don't think we ever should actually hit the line 137 and that whole clause of `match`. Because if we got to the Leaf, then we know that this is THE node, so line 127 should return `false` and we should return `Ok(false)` with incomplete key. 
But, I also don't think that completeness of the key maters. What maters is that it is a key prefix that defines the same set of vertices as a full key -- only the Leaf node. Slice of key in the Leaf node is kinda redundant suffix in terms of the trie structure. 
My *intuition* here is that we only use this key_nibbles to iterate trie -- not to retrieve nodes (needs checking), and we iterate trie based on function that works with prefixes (again, needs checking, I'm talking about `seek_nibble_slice`) 
In slightly different words, I think we only working in Trie based on key comparison as in <=, >=, not equality. All of these inequalities will yield the same value if we switch full leaf key for that truncated leaf key, but only if we are only working with this Trie elements. That is true because we constructed such a key prefix that is long enough to include at least one different character from every other Trie element.